tensorflow/python/framework/device_spec.py

```diff
@@ -302,6 +302,15 @@ def _get_combined_properties(self, dev):
         dev.device_index if dev.device_index is not None else self.device_index,
     )
 
+  @staticmethod
+  def _get_valid_device_types():
+    valid_device_types = set({})
+    physical_devices = pywrap_tfe.TF_ListPluggablePhysicalDevices()
+    for device in physical_devices:
+      valid_device_types.add(device.decode().split(":")[1]) 
```

Comment:
Can this set of valid devices be cached and re-generated only on failure? There may be quite a few physical devices, and parsing device strings happens quite frequently.

Reply:
`_get_valid_device_type()` is only called by _string_to_components, I think device_spec is already cached by 
```

327     cached_result = _STRING_TO_COMPONENTS_CACHE.get(spec)
328     if cached_result is not None:
329       return cached_result
```

`_get_valid_device_type` itself can not do cache because it's first time called  is earlier than `load_pluggable_device_library`, so it can not get the plug in device type if it is cached.